### PR TITLE
Adding Msft ALS Recommender Helpers

### PR DESCRIPTION
**Pull request by dciborow**

### CODE REVIEW

**Comment by dciborow** [comment]
restarting pull request with cleaner commit history

**Comment by mhamilton723** [comment]
Hey, thanks for your hard work! Please just ping me when you need a review of a working build!

File: notebooks/tests/MsftRecommendations.ipynb [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+                 "cells": [
+                  {
+                   "cell_type": "code",
+                   "execution_count": null,
+                   "metadata": {
+                    "collapsed": true
+                   },
+                   "outputs": [],
+                   "source": [
+                    "# Use the Azure Machine Learning data collector to log various metrics\r\n",
```
Comment (drdarshan) [inline notebooks/tests/MsftRecommendations.ipynb]: Could you please remove the data collection code from this notebook? It's intended to work on all Spark clusters.

File: src/msft-recommendation/src/main/scala/TrainValidRecommendSplit.scala [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+
+    Array(trainingDataset, validationDataset)
+  }
+
+  def prepareTestData(validationDataset: DataFrame, recs: DataFrame): Dataset[_] = {
+    import org.apache.spark.sql.functions.{collect_list, rank => r}
+
+    val est = $(estimator) match {
+      case p: Pipeline => {
+        //Assume Rec is last stage of pipeline
```
Comment (drdarshan) [inline src/msft-recommendation/src/main/scala/TrainValidRecommendSplit.scala]: This is a somewhat unusual requirement - what is the use-case for passing in a pipeline as the estimator here?

File: src/msft-recommendation/src/test/scala/MsftRecommendationSpec.scala [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+      .setNumUserBlocks(200)
+      .setNumItemBlocks(20)
+      .setItemCol("itemID")
+      .setUserCol("customerID")
+      .setRatingCol("rating")
+      .fit(dfFit)
+
+    val items = model.recommendForAllUsers(3)
+    val users = model.recommendForAllItems(3)
+    //    val predictions = model.transform(dfTransform)
```
Comment (drdarshan) [inline src/msft-recommendation/src/test/scala/MsftRecommendationSpec.scala]: Minor: please remove commented-out code.

File: src/msft-recommendation/src/test/scala/TrainValidRecommendSplitRec.scala [hunk 4]
```diff
@@ -0,0 +1,596 @@
...
+
+    List(new TestObject(tvRecommendationSplit, transformedDf))
+  }
+
+  override def reader: MLReadable[_] = TrainValidRecommendSplit
+
+  override def modelReader: MLReadable[_] = TrainValidRecommendSplitModel
+
+//  test("testALSPipeline") {
+//    val dfRaw2: DataFrame = session
```
Comment (drdarshan) [inline src/msft-recommendation/src/test/scala/TrainValidRecommendSplitRec.scala]: Maybe remove this? 
Alternatively for large dataset, you can move it to a text file and parse it inside the testcase.

File: src/msft-recommendation/src/test/scala/TrainValidRecommendSplitRec.scala [hunk 5]
```diff
@@ -0,0 +1,596 @@
...
+      .setTrainRatio(0.8)
+      .setUserCol(customerIndex.getOutputCol)
+      .setRatingCol("rating")
+      .setItemCol(ratingsIndex.getOutputCol)
+
+    val tvModel = tvRecommendationSplit.fit(transformedDf)
+
+    val model = tvModel.bestModel.asInstanceOf[MsftRecommendationModel]
+
+    val items = model.recommendForAllUsers(3)
```
Comment (drdarshan) [inline src/msft-recommendation/src/test/scala/TrainValidRecommendSplitRec.scala]: I don't see any asserts - is this checking for something not throwing?

**Pull request closed without merging**